### ISSUE TITLE
pyright fails to determine variable type correctly

### ISSUE THREAD

**Issue by lovetox** [microsoft/pyright#2288]
**Describe the bug**
pyright is not able to determine the type correctly it raises

```
error: Type of "ns" is partially unknown
    Type of "ns" is "Unknown | str" (reportUnknownVariableType)
```

**To Reproduce**
run the code with typeCheckingMode:  strict

**Expected behavior**
pyright should see that `ns` can never be something different than a string and should not raise this error

**Screenshots or Code**
```python
def test(parts: List[str]):
    ns = ''
    for part in parts:
        if ns:
            ns += f'.{part}'
        else:
            ns += part
```

**VS Code extension or command-line**
command-line version 1.1.166

**Comment by erictraut** [comment]
Thanks, I'm able to repro the problem. Here's an even simpler version:

```python
# pyright: strict

from typing import Iterable

def test(parts: Iterable[str]):
    ns = ""
    for _ in parts:
        if ns:
            ns = ns + ""
```

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.167, which I just published. It will also be included in the next release of pylance.

**Comment by lovetox** [comment]
This does still not work fully correct

if i use `ns` later its `str | Unkown`

```python
def build_namespaces(namespace: str) -> List[str]:
    parts = namespace.split('.')
    namespaces: List[str] = []
    ns = ''
    for part in parts:
        if ns:
            ns += f'.{part}'
        else:
            ns += part
        namespaces.append(ns)
    return namespaces
```

Reports
```
  helpers.py:66:27 - error: Argument type is partially unknown
    Argument corresponds to parameter "__object" in function "append"
    Argument type is "str | Unknown" (reportUnknownArgumentType)
```

Do you want me to open a new issue?

**Comment by erictraut** [comment]
Thanks, I just opened a new issue.

**Comment by erictraut** [comment]
You probably already realized this, but you can easily work around this by adding a `str` type annotation to `ns`.

```python
    ns: str = ''
```

**Comment by lovetox** [comment]
no i did not, thanks.
